File: serenity-app/src/test/java/us/nineworlds/serenity/GalleryOnItemClickListenerTest.java
package us.nineworlds.serenity;

import android.view.View;
import org.assertj.android.api.Assertions;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.Shadows;
import org.robolectric.shadows.ShadowActivity;
import us.nineworlds.serenity.core.menus.MenuItem;
import us.nineworlds.serenity.ui.browser.movie.MovieBrowserActivity;
import us.nineworlds.serenity.ui.browser.tv.TVShowBrowserActivity;
import us.nineworlds.serenity.ui.home.MainMenuTextViewAdapter;
import us.nineworlds.serenity.ui.preferences.LeanbackSettingsActivity;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.doReturn;
import static org.mockito.MockitoAnnotations.initMocks;

@RunWith(RobolectricTestRunner.class)
public class GalleryOnItemClickListenerTest {

  @Mock
  MainMenuTextViewAdapter mockAdapter;

  MainMenuTestActivity context;
  GalleryOnItemClickListener onItemClickListener;

  @Before
  public void setUp() throws Exception {
    initMocks(this);
    context = Robolectric.setupActivity(MainMenuTestActivity.class);
    onItemClickListener = new GalleryOnItemClickListener();
  }

  @After
  public void tearDown() {
    context.finish();
  }

  @Test
  public void onClickMenuSearchLaunchesExpectedActivity() {
    View view = new View(context);
    MenuItem searchMenuItem = new MenuItem();
    searchMenuItem.setType("search");

    onItemClickListener.onItemClick(view, searchMenuItem);

    assertThat(context.onSearchActivtyCalled).isTrue();
  }

  @Test
  public void onClickOpensOptionsMenu() {
    View view = new View(context);
    MenuItem searchMenuItem = new MenuItem();
    searchMenuItem.setType("options");

    onItemClickListener.onItemClick(view, searchMenuItem);

    assertThat(context.openOptionsMenu).isTrue();
  }

  @Test
  public void onClickLaunchesMovieBrowserActivity() {
    View view = new View(context);
    MenuItem searchMenuItem = new MenuItem();
    searchMenuItem.setType("movie");

    onItemClickListener.onItemClick(view, searchMenuItem);

    ShadowActivity shadowActivity = Shadows.shadowOf(context);
    Assertions.assertThat(shadowActivity.getNextStartedActivity()).hasComponent(context, MovieBrowserActivity.class);
  }

  @Test
  public void onClickLaunchesTVShowBrowserActivity() {
    View view = new View(context);
    MenuItem searchMenuItem = new MenuItem();
    searchMenuItem.setType("show");

    onItemClickListener.onItemClick(view, searchMenuItem);

    ShadowActivity shadowActivity = Shadows.shadowOf(context);
    Assertions.assertThat(shadowActivity.getNextStartedActivity()).hasComponent(context, TVShowBrowserActivity.class);
  }

  @Test
  public void onClickUnknownRequestLaunchesPreferences() {
    View view = new View(context);
    MenuItem searchMenuItem = new MenuItem();
    searchMenuItem.setType("unknown");

    onItemClickListener.onItemClick(view, searchMenuItem);

    ShadowActivity shadowActivity = Shadows.shadowOf(context);
    Assertions.assertThat(shadowActivity.getNextStartedActivity())
        .hasComponent(context, LeanbackSettingsActivity.class);
  }
}

File: serenity-app/src/test/kotlin/us/nineworlds/serenity/GalleryOnItemClickListenerTest.kt
package us.nineworlds.serenity

import android.content.Intent
import android.view.View
import assertk.assertThat
import assertk.assertions.isEqualTo
import assertk.assertions.isTrue
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.mockk
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows
import us.nineworlds.serenity.core.menus.MenuItem
import us.nineworlds.serenity.ui.preferences.LeanbackSettingsActivity

@RunWith(RobolectricTestRunner::class)
class GalleryOnItemClickListenerTest {

  private val mockAdapter: MainMenuTextViewAdapter = mockk(relaxed = true)

  private lateinit var context: MainMenuTestActivity
  private lateinit var onItemClickListener: GalleryOnItemClickListener

  @Before
  fun setUp() {
    context = Robolectric.setupActivity(MainMenuTestActivity::class.java)
    onItemClickListener = GalleryOnItemClickListener(mockAdapter)
  }

  @After
  fun tearDown() {
    context.finish()
    clearAllMocks()
  }

  @Test
  fun onClickMenuSearchLaunchesExpectedActivity() {
    val view = View(context)
    val searchMenuItem = MenuItem().apply {
      type = "search"
    }
    every { mockAdapter.getItemAtPosition(any()) } returns searchMenuItem

    onItemClickListener.onItemClick(view, 0)

    assertThat(context.onSearchActivtyCalled).isTrue()
  }

  @Test
  fun onClickOpensOptionsMenu() {
    val view = View(context)
    val optionsMenuItem = MenuItem().apply {
      type = "options"
    }
    every { mockAdapter.getItemAtPosition(any()) } returns optionsMenuItem

    onItemClickListener.onItemClick(view, 0)

    assertThat(context.openOptionsMenu).isTrue()
  }

  @Test
  fun onClickUnknownRequestLaunchesPreferences() {
    val view = View(context)
    val unknownMenuItem = MenuItem().apply {
      type = "unknown"
    }
    every { mockAdapter.getItemAtPosition(any()) } returns unknownMenuItem

    onItemClickListener.onItemClick(view, 0)

    val shadowActivity = Shadows.shadowOf(context)
    val nextStartedActivity = shadowActivity.nextStartedActivity

    val expectedIntent = Intent(context, LeanbackSettingsActivity::class.java)
    assertThat(nextStartedActivity.component).isEqualTo(expectedIntent.component)
  }
}